Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool.swift

```diff
@@ -121,6 +128,376 @@ enum HTTPConnectionPool {
+            self.scheduleIdleTimerForConnection(connectionID, on: eventLoop)
+
+        case .cancelTimeoutTimer(let connectionID):
+            self.cancelIdleTimerForConnection(connectionID)
+
+        case .closeConnection(let connection, isShutdown: let isShutdown):
+            // we are not interested in the close future...
+            connection.close(promise: nil)
+
+            if case .yes(let unclean) = isShutdown {
+                self.delegate.connectionPoolDidShutdown(self, unclean: unclean)
+            }
+
+        case .cleanupConnections(let cleanupContext, isShutdown: let isShutdown):
+            for connection in cleanupContext.close {
+                connection.close(promise: nil)
+            }
+
+            for connection in cleanupContext.cancel {
+                connection.close(promise: nil)
+            }
+
+            for connectionID in cleanupContext.connectBackoff {
+                self.cancelConnectionStartBackoffTimer(connectionID)
+            }
+
+            if case .yes(let unclean) = isShutdown {
+                self.delegate.connectionPoolDidShutdown(self, unclean: unclean)
+            }
+
+        case .none:
+            break
+        }
+    }
+
+    private func runRequestAction(_ action: StateMachine.RequestAction) {
+        // The order of execution fail/execute request vs cancelling the request timeout timer does
+        // not matter in the actions here. The actions don't cause any side effects that will be
+        // reported back to the state machine and are not dependent on each other.
+
+        switch action {
+        case .executeRequest(let request, let connection, cancelTimeout: let cancelTimeout):
+            if cancelTimeout {
+                self.cancelRequestTimeout(request.id)
+            }
+            connection.executeRequest(request.req)
+
+        case .executeRequestsAndCancelTimeouts(let requests, let connection):
+            self.cancelRequestTimeouts(requests)
+            requests.forEach { connection.executeRequest($0.req) }
+
+        case .failRequest(let request, let error, cancelTimeout: let cancelTimeout):
+            if cancelTimeout {
+                self.cancelRequestTimeout(request.id)
+            }
+            request.req.fail(error)
+
+        case .failRequestsAndCancelTimeouts(let requests, let error):
+            self.cancelRequestTimeouts(requests)
+            requests.forEach { $0.req.fail(error) }
+
+        case .scheduleRequestTimeout(let request, on: let eventLoop):
+            self.scheduleRequestTimeout(request, on: eventLoop)
+
+        case .cancelRequestTimeout(let requestID):
+            self.cancelRequestTimeout(requestID)
+
+        case .none:
+            break
+        }
+    }
+
+    private func createConnection(_ connectionID: Connection.ID, on eventLoop: EventLoop) {
+        // Even though this function is called make it actually creates/establishes a connection.
+        // TBD: Should we rename it? To what?
```

Comment:
Maybe just changing "make" to "create" is enough to differentiate make it clear that it's not a factory. Alternatively something like `provideNewConnection(to: self, ...)`?